src/main/resources/include/hail/NativeObj.h

```diff
@@ -0,0 +1,52 @@
+#ifndef HAIL_NATIVEOBJ_H
+#define HAIL_NATIVEOBJ_H 1
+
+#include <memory>
+
+// Declare a function to be exported from a DLL with "C" linkage
+
+#define NATIVEMETHOD(cppReturnType, scalaClass, scalaMethod) \
+  extern "C" __attribute__((visibility("default"))) \
+    cppReturnType Java_is_hail_nativecode_##scalaClass##_##scalaMethod
+
+namespace hail {
+
+// The lifetime of off-heap objects may be controlled by any combination
+// of C++ std::shared_ptr<T> and Jvm-side NativePtr objects.  But classes
+// used in this way must inherit from NativeObj.
+
+class NativeObj :
+  public std::enable_shared_from_this<NativeObj> {
+  public:
+    // Objects managed with a Jvm-side NativePtr need to be destroyed
+    // without knowing their precise type, so they must have a virtual
+    // destructor.
+    virtual ~NativeObj() { }
+
+    // Subclasses should override getClassName, which may be helpful
+    // for debugging.
+    virtual const char* get_class_name() { return "NativeObj"; }
+    
+    // If a subclass wants to make some fields visible to Jvm Unsafe
+    // access, it can publish their offsets.  This is probably only
+    // useful for integer types, so we pass in one parameter to
+    // identify the fieldSize as 1, 2, 4, or 8 bytes.
+    virtual long get_field_offset(int fieldSize, const char* fieldName) { return(-1); }
+};
+
+// On the Jvm side, we don't have distinct classes/types for NativePtr
+// to different off-heap objects - we treat them all as NativeObjPtr.
+typedef std::shared_ptr<NativeObj> NativeObjPtr;
+
+#define MAKE_NATIVE(ObjT, ...) \
+  (std::static_pointer_cast<NativeObj>(std::make_shared<ObjT>(__VA_ARGS__)))
```

Comment:
I don't think there's any reason to avoid the standard C++11 perfect forwarding pattern, e.g.
```C++
template<typename T, typename... Ts>
std::shared_ptr<NativeObj> make_native(Ts&&... params) {
  return std::make_shared<T>(std::forward<Ts>(params))
}
```
though I'm not sure if `make_native<ObjT>(args...)` is much improvement over `make_shared<ObjT>(args...)` implicitly converted to a `shared_ptr<NativeObj>`.

Reply:
Does that syntax actually work ?  Because when I look at std::make_shared<> there are some
much more complicated things going on, I think perhaps to do with unpacking the multiple
args with a recursive template and doing std::forward on each one ?

Reply:
I got that from Meyers, Effective Modern C++ page 139. Though I made a typo, it should be
```C++
template<typename T, typename... Ts>
std::shared_ptr<NativeObj> make_native(Ts&&... params) {
  return std::make_shared<T>(std::forward<Ts>(params)...)
}
```
You can see the same pattern at http://en.cppreference.com/w/cpp/utility/forward. I'll try it out to verify.

The recursive template you're seeing is maybe to be compatible with compilers not supporting variadic templates (introduced in C++11)?

Reply:
This works for me, using clang++ -std:c++11:
```C++
#include<memory>
#include<vector>
#include<iostream>

template<typename T, typename... Ts>
std::shared_ptr<T> make_shared(Ts&&... params) {
  return std::shared_ptr<T>(new T(std::forward<Ts>(params)...));
}

int main(int argc, char *argv[]) {
  auto vec = make_shared<std::vector<int>>(10, 0);
  for (const int& i : *vec)
    std::cout << i << ' ';
}

```

Reply:
Oh, I see.  There's some magic to do with "parameter pack expansion" which makes
the (std::forward<Ts>(params)...) get expanded out into the equivalent of a
comma-separated list of std::forward<T0>(param0), std::forward<T1>(param1), 

I'll do it that way since it's approved modern C++ style, but to my mind they've
built an excessively complex and hard-to-understand new way of doing something
that worked just fine as a macro :-(

Reply:
General comment - you can write a lot of C++ without knowing the corners of the
language standard :-)

Reply:
Sure, and I do apologize if I'm being a language lawyer. My goal isn't to know every dark corner of the standard (I think I'd go crazy), I'm just trying to find my way to the current community consensus of best practices.

Reply:
Hey, I'm learning something new, and that's good.  And your comments have allowed me to
get rid of MAKE_NATIVE completely, which is definitely better.